File: PDOO/Practica4/Ruby/Irrgarten/player.rb
require_relative 'dice'
require_relative 'weapon'
require_relative 'shield'
require_relative 'directions'
require_relative 'labyrinth_character'

module Irrgarten
  class Player < LabyrinthCharacter
    @@MAX_WEAPONS = 2
    @@MAX_SHIELDS = 3
    @@INITIAL_HEALTH = 10
    @@HITS2LOSE = 3

    def initialize(number, intelligence, strength)
      super("Player ##{number}", intelligence, strength, @@INITIAL_HEALTH)
      @weapons = []
      @shields = []
      @number = number

      @row = 0
      @col = 0

      @consecutive_hits = 0
    end

    # TODO: Constructor de copia?

    # Resurrects the player, setting its health to the initial value provided by
    # the constant @@INITIAL_HEALTH, and clearing its weapons and shields
    # @return [void]
    def resurrect
      @health = @@INITIAL_HEALTH
      @weapons.clear
      @shields.clear
      reset_hits
    end

    attr_reader :number

    # Moves the player in the preferred direction if it is valid, or in the first
    # valid direction if the preferred one is not valid
    # @param [Direction] direction The preferred direction to move
    # @param [Array<Direction>] valid_moves The valid directions to move
    # @return [Direction] The direction the player will move
    def move(direction, valid_moves)
      size = valid_moves.size
      contained = valid_moves.include?(direction)

      return valid_moves[0] if size.positive? && !contained

      direction
    end

    # Returns the attack power of the player based on its strength and the sum of
    # its weapons
    # @return [Float] The attack power of the player
    def attack
      @strength + sum_weapons
    end

    # Defends the player from an attack, reducing its health if the attack is successful
    # and returning whether the player is dead or not
    # @param [Float] received_attack The attack power received by the player
    # @return [Boolean] True if the player is dead, false otherwise
    def defend(received_attack)
      manage_hit(received_attack)
    end

    # Receives a reward after winning a combat, consisting of a random number of weapons,
    # shields, and health units based on the constants defined in the Dice class
    # @return [void]
    def receive_reward
      w_reward = Dice.weapons_reward
      s_reward = Dice.shields_reward

      w_reward.times do
        receive_weapon(new_weapon)
      end

      s_reward.times do
        receive_shield(new_shield)
      end

      @health += Dice.health_reward
    end

    # Returns a string representation of the player, including its name, health, intelligence, strength,
    # and the number of consecutive hits it has received, as well as its weapons and shields
    # @return [String] The string representation of the player
    def to_s
      base_str = super.chomp(" \n")

      weapons_str = @weapons.map(&:to_s).join(' ')
      shields_str = @shields.map(&:to_s).join(' ')
      additional_info = format(", %i", @consecutive_hits)

      "#{base_str}#{additional_info}] \n\tWeapons: #{weapons_str}\n\tShields: #{shields_str}"
    end

    protected

    # Returns the sum of the attack power of the player's weapons
    # @return [Float] The sum of the attack power of the player's weapons
    def sum_weapons
      sum = 0.0
      @weapons.each do |weapon|
        sum += weapon.attack
      end
      sum
    end

    # Returns the sum of the protection power of the player's shields
    # @return [Float] The sum of the protection power of the player's shields
    def sum_shields
      sum = 0.0
      @shields.each do |shield|
        sum += shield.protect
      end
      sum
    end

    # Returns the defensive energy of the player, which is the sum of its intelligence
    # and the sum of the protection power of its shields
    # @return [Float] The defensive energy of the player
    def defensive_energy
      @intelligence + sum_shields
    end

    private

    # Receives a weapon w, it discards some weapons based on a probability and
    # adds the weapon to the player's weapons if the player has less than the maximum
    # number of weapons allowed
    # @param [Weapon] w The weapon to be received
    # @return [void]
    def receive_weapon(w)
      @weapons.delete_if(&:discard)

      return unless @weapons.size < @@MAX_WEAPONS

      @weapons.push(w)
    end

    # Receives a shield s, it discards some shields based on a probability and
    # adds the shield to the player's shields if the player has less than the maximum
    # number of shields allowed
    # @param [Shield] s The shield to be received
    # @return [void]
    def receive_shield(s)
      @shields.delete_if(&:discard)

      return unless @shields.size < @@MAX_SHIELDS

      @shields.push(s)
    end

    # Returns a new weapon with random power and uses left
    # @return [Weapon] The new weapon
    def new_weapon
      Weapon.new(Dice.weapon_power, Dice.uses_left)
    end

    # Returns a new shield with random power and uses left
    # @return [Shield] The new shield
    def new_shield
      Shield.new(Dice.shield_power, Dice.uses_left)
    end

    # Manages the hit received by the player, reducing its health if the attack is successful
    # and returning whether the player is dead or not
    # @param [Float] received_attack The attack power received by the player
    # @return [Boolean] True if the player is dead, false otherwise
    def manage_hit(received_attack)
      if defensive_energy < received_attack
        got_wounded
        inc_consecutive_hits
      else
        reset_hits
      end

      if @consecutive_hits == @@HITS2LOSE || dead
        reset_hits
        return true
      end

      false
    end

    # Resets the number of consecutive hits received by the player
    # @return [void]
    def reset_hits
      @consecutive_hits = 0
    end

    # Increases the number of consecutive hits received by the player
    # @return [void]
    def inc_consecutive_hits
      @consecutive_hits += 1
    end
  end
end

File: PDOO/Practica4/Java/Irrgarten/src/irrgarten/Player.java
/**
 * @file Player.java
 * @author Miguel Angel De la Vega Rodríguez
 * @author Alberto De la Vera Sanchez
 * @description This file contains the Player class, which provides methods to
 * manage the player.
 */

package irrgarten;

import java.util.ArrayList;


public class Player extends LabyrinthCharacter{
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;

    private char number;
    private int consecutiveHits;
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;

    public Player(char number, float intelligence, float strength) {
        super("Player #", number, intelligence, strength);
        this.consecutiveHits = 0;
        this.number = number;

        this.weapons = new ArrayList<Weapon>();
        this.shields = new ArrayList<Shield>();
    }

    /**
     * It resurrects the player, setting its health to the initial value,
     * resetting the number of consecutive hits and setting the player's weapons
     * and shields to empty lists
     * 
     */
    public void resurrect() {
        this.setHealth(INITIAL_HEALTH);
        this.weapons.clear();
        this.shields.clear();
        resetHits();
    }

    /**
     * Returns the number of the player
     * 
     * @return number
     */
    public char getNumber() {
        return number;
    }

    /**
     * Returns the direction of the move if the move is possible and 
     * the first move possible in validMoves otherwise
     * 
     * @param direction Preferred direction to move
     * @param validMoves Valid moves array
     * @return The direction of the possible move 
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves) {
        if (validMoves.size() > 0 && !validMoves.contains(direction)) {
            return validMoves.get(0);
        }
        else{
            return direction;
        }
    }

    /**
     * Calculates the strength of the player and the sum of the power provided
     * by its weapons
     *
     * @return the attack value
     */
    public float attack() {
        return sumWeapons()+this.getStrength();
    }

    /**
     * This method delegates its functionality to the manageHit method
     *
     * @param receivedAttack the attack value received
     * @return manageHit(receivedAttack)
     */
    public boolean defend(float receivedAttack) {
        return manageHit(receivedAttack);
    }

    /**
     * This method creates a reward for the player using random possibility
     * using Dice, it rewards the player with weapons, shields and health
     */
    public void receiveReward() {
        int wReward = Dice.weaponsReward();
        int sReward = Dice.weaponsReward();

        for (int i = 1; i <= wReward; ++i) {
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }

        for (int i = 1; i <= sReward; ++i) {
            Shield snew = newShield();
            receiveShield(snew);
        }

        this.setHealth(this.getHealth() + Dice.healthReward());
    }

    /**
     * Returns a string with the current state of the player, with the format:
     * P[number, health, intelligence, strength]
     *      Weapons: Weapon1    Weapon2 ... WeaponN
     *      Shields: Shield1    Shield2 ... ShieldN
     *
     * @return Current state of the player
     */
    public String toString() {
        String toReturn = "P" + super.toString();

        toReturn += "\tWeapons: ";
        for (Weapon weapon : weapons) {
            toReturn += weapon.toString() + "\t";
        }

        toReturn += "\n\tShields: ";
        for (Shield shield : shields) {
            toReturn += shield.toString() + "\t";
        }

        return toReturn;
    }

    /**
     * This method updates the weapons of the player if any has to be
     * discarded and adds one weapon if MAX_WEAPONS is not exceeded
     * 
     * @param w New weapon to be added
     */
    private void receiveWeapon(Weapon w) {
        for (int i = 0; i < weapons.size(); ++i) {
            if (weapons.get(i).discard()) {
                weapons.remove(i);
                --i;
            }
        }

        if (weapons.size() < MAX_WEAPONS) {
            weapons.add(w);
        }
    }

    /**
     * This method updates the shields of the player if any has to be 
     * discarded and adds one shield if MAX_SHIELDS is not exceeded
     * 
     * @param s New shield to be added
     */
    private void receiveShield(Shield s) {
        for (int i = 0; i < shields.size(); ++i){
            if (shields.get(i).discard()){
                shields.remove(i);
                --i;
            }
        }

        if (shields.size() < MAX_SHIELDS) {
            shields.add(s);
        }
    }

    /**
     * Creates a new instance of the Weapon class and adds it, parameters needed to
     * generate the weapon are taken from Dice class
     *
     * @return new weapon
     */
    private Weapon newWeapon() {
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }

    /**
     * Creates a new instance of the Shield class and adds it, parameters needed to
     * generate the shield are taken from Dice class
     *
     * @return new shield
     */
    private Shield newShield() {
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }

    /**
     * Returns the sum of the attack values of the weapons calling attack method
     *
     * @return sum of the attack values
     */
    protected float sumWeapons() {
        float sum = 0f;
        for (Weapon weapon : weapons) {
            sum += weapon.attack();
        }
        return sum;
    }

    /**
     * Returns the sum of the protection provided by the shields calling
     * protect method
     *
     * @return sum of the protection
     */
    protected float sumShields() {
        float sum = 0f;
        for (Shield shield : shields) {
            sum += shield.protect();
        }
        return sum;
    }

    /**
     * Returns the sum of the intelligence and the protection provided by the
     * shields
     *
     * @return the sum 
     */
    protected float defensiveEnergy() {
        return sumShields() + getIntelligence();
    }

    /**
     * It manages the attack received by the player, if the attack is greater
     * than the defensive energy of the player, the player is wounded and the number
     * of consecutive hits is incremented, otherwise the number of consecutive hits
     * is reset
     * 
     * @param receivedAttack power of the received attack
     * @return whether the player loses the combat or not
     */
    private boolean manageHit(float receivedAttack) {
        if (defensiveEnergy() < receivedAttack) {
            gotWounded();
            incConsecutiveHits();
        }
        else{
            resetHits();
        }

        boolean lose = (consecutiveHits == HITS2LOSE) || dead();

        if (lose) {
            resetHits();
        }

        return lose;
    }

    /**
     * Resets the number of consecutive hits to 0
     */
    private void resetHits() {
        consecutiveHits = 0;
    }

    /**
     * Increments the number of consecutive hits by one unit
     */
    private void incConsecutiveHits() {
        ++consecutiveHits;
    }

}
